arch/lkl/kernel/syscalls.c

```diff
@@ -296,10 +298,12 @@ long lkl_syscall(long no, long *params)
 
 out:
 	lkl_cpu_put();
-
+	
 	LKL_TRACE("done (no=%li task=%s current=%s ret=%i)\n", no,
 		  task ? task->comm : "NULL", current->comm, ret);
-
+	/* once cpu is released, run any pending signal handlers */
+	/* experiment indicates that all the signal sending happens here, not in the idle loop. */ 
+	send_current_signals(NULL);
```

Comment:
Ideally, we'd set up a plausible ucontext here.  If we had a return from `move_signals_to_task` telling us if there are any pending signals for this task, we could try to set up a plausible return frame here.  Ideally, we'd actually store all callee-save registers before entry to `lkl_syscall` and pass them in as an argument (or keep the existing `lkl_syscall` and expose an `lkl_syscall_and_deliver_signals` that would take a ucontext as an extra argument).

Reply:
Maybe I could copy the context as part of the list node. Then it will be available to the signal sending code. Everyone previously calling do_signal really had no context, passing NULL. I'm not sure where to find the context. It it the one from when you raise a signal? Will it really just be whatever lib pthread had hanging about?

Reply:
When a signal is delivered during a syscall, it should get the ucontext / mcontext that the kernel saved on entry to the syscall and this should then be restored during sigreturn.  Unfortunately, because our syscalls are just function calls, we don't save a context.  We should probably wrap lkl_syscall in an assembly routine that allocates a ucontext on the caller stack and saves all callee-save registers into it (temporary / argument registers, including all of the x87 / AVX frame, can be have undefined contents).

Reply:
Still to do.

Reply:
To do.